IndexedCP Python library

adithya1012 opened the pull request:
## Python Implementation of IndexedCP

Short file upload with Encryption: https://youtube.com/shorts/00Lj3eQnLVs
Old One wo Encryption Short file upload: https://www.youtube.com/shorts/RSIwcoRnTu0

### Summary
Complete Python port of IndexedCP with **end-to-end encryption support**. Implements RSA-4096 + AES-256-GCM encryption compatible with Node.js server and encrypted storage.

###  What's New

**Encryption Support**
- **RSA-4096** key wrapping + **AES-256-GCM** data encryption
- **Encrypted storage** at rest (JSON-based, matching JavaScript implementation)
- **Keystore abstraction** (filesystem backend with file permissions)
- **Per-stream ephemeral keys** for secure file uploads
- **Public key caching** for offline encryption
- **Backward compatible** - encryption is optional

**Core Features**
- **Client**: Chunked uploads with SQLite/encrypted storage backends
- **Crypto Utils**: Low-level cryptographic primitives
- **Test Suite**: 82 passing tests (15 integration tests require Node.js server)
- **Interactive Demos**: Server and client demos with encryption
- **Documentation**: Encryption guides and API references

### Key Features

**Encryption Architecture**
- **Envelope encryption**: RSA wraps AES session keys
- **Unique session keys**: Per-file encryption isolation

**Client Features**
- Chunked uploads with encryption
- Offline encryption capability
- Retry logic with exponential backoff
- Progress callbacks
- Background upload mode

**Testing**
- 82/82 tests passing
- Full encryption workflow coverage
- Backward compatibility verified
- Integration tests 

### Interactive Demos

1. **Server Demo** (`examples/server_demo.py`):
   - Displays configuration
   - Real-time upload logging
   - Clean shutdown handling

2. **Client Demo** (`examples/client_demo.py`):
   - Simple file upload
   - Batch uploads
   - Large file chunking with DB inspection
   - Background upload mode
   - Automatic cleanup

### Issues 

https://github.com/adithya1012/IndexedCP/issues/12 -> 3 Sub issues  
https://github.com/adithya1012/IndexedCP/issues/16 -> 4 Sub issues
https://github.com/adithya1012/IndexedCP/issues/21 -> 4 Sub issues

adithya1012 committed:
feat(python): add Python project structure and build configuration.

- Initialize Python implementation of IndexedCP with minimal project setup.
- Add colored logger module with comprehensive tests

horner commented on python/tests/test_basic.py:
```diff
@@ -0,0 +1,21 @@
+@pytest.mark.asyncio
+async def test_async_support():
+    """Test that async tests work"""
+    assert True
```
?? They dont test anything

adithya1012 replied:
Removed the test template functions. Written 3 comprehensive test files for client, server and integration.

horner commented on python/README.md:
```diff
@@ -0,0 +1,71 @@
+
+## Features
+
+- 🚀 **Minimal Code**: Simple, focused implementation
```
You should take fluff like this out of there. Also be careful when you ask for minimal code that it doesn’t just give you a test with nothing in it.

adithya1012 replied:
Updated the code by removing emojis (🚀 📋 ✅ 🌐 💡 🛑 📝 💾 ⏸️ ⏳ 📤 🔧 🔌 🎬 🎉 📁 ⚠️ 🧹 📊)

horner commented on python/README.md:
```diff
@@ -0,0 +1,71 @@
+## Features
+
+- 🚀 **Minimal Code**: Simple, focused implementation
+- 📦 **Easy Setup**: Standard Python packaging
```
Try not to waste your readers time. One of the things I hate about AI is that it puts a lot of extra fluff in sometimes.

horner replied:
Self promotional fluff

adithya1012 replied:
Updated the PR with 4 Commits.

horner reviewed (changes requested):
This commit is really not worthy of anything yet. Call it WIP still. It is the structure and the foundation on which to build.  It’s always good to do a first commit like this and build on it. I’ll merge when we meet the goals.

adithya1012 committed:
feat: add pluggable storage abstraction with built-in SQLite backend
Implement zero-dependency storage layer for Python IndexedCP following
JavaScript keystore pattern. Uses Python's built-in sqlite3 module for
persistence without external database dependencies.

adithya1012 committed:
Add Python implementation of IndexedCP client with storage abstraction layer
Implemented IndexedCPClient with file chunking, SQLite storage backend, and HTTP upload queue. Created storage abstraction (BaseStorage, SQLiteStorage) with async context manager support. Added comprehensive test suite (44 tests) covering client operations, storage layer, and logging utilities. Includes examples and documentation following JS implementation patterns.

adithya1012 committed:
feat(python): add upload functionality with retry logic and progress tracking

- Implement upload_buffered_files() with chunked uploads
- Add exponential backoff retry logic (configurable)
- Add progress callbacks (on_upload_progress, on_upload_error, on_upload_complete)
- Add background upload process (non-blocking)
- Add session tracking and automatic cleanup

adithya1012 committed:
feat(python): implement IndexedCP server with chunked upload support

- Add IndexedCPServer class with FastAPI backend for file uploads
- Implement three path handling modes: ignore (default), sanitize, allow-paths
- Add chunked upload reassembly with session tracking
- Include API key authentication middleware
- Add integration tests and interactive demos

adithya1012 commented on python/README.md:
```diff
@@ -0,0 +1,71 @@
+
+- 🚀 **Minimal Code**: Simple, focused implementation
+- 📦 **Easy Setup**: Standard Python packaging
+
```
Need this library for future

adithya1012 commented on python/tests/test_basic.py:
```diff
@@ -0,0 +1,21 @@
+
+def test_basic_assertion():
+    """Test that pytest is working correctly"""
+    assert True
```
Removed Basic test templates

adithya1012 commented on python/pyproject.toml:
```diff
@@ -0,0 +1,55 @@
+    "pytest-asyncio>=0.21.0",
+    "pytest-cov>=4.1.0",
+    "black>=23.0.0",
+    "ruff>=0.1.0",
```
Updated

adithya1012 commented on python/tests/test_basic.py:
```diff
@@ -0,0 +1,21 @@
+@pytest.mark.asyncio
+async def test_async_support():
+    """Test that async tests work"""
+    assert True
```
Removed the test template functions

horner commented on python/README.md:
There is A lot in here that is too noisy. I would just do the TLDR and the details on how to use it. Not how it works. 


I also think it should work like the node layer works. The files paths for be the same.

adithya1012 replied:
Reduced ReadMe content to 153 lines

Both client implementations use same path:
Client DB: ~/.indexcp/db/client.db (user's home directory) 
[Node](https://github.com/adithya1012/IndexedCP/blob/a549b5be3a0f5a451e53fd57736b81c07eb271ba/lib/client.js#L37) [Python](https://github.com/adithya1012/IndexedCP/blob/a549b5be3a0f5a451e53fd57736b81c07eb271ba/python/src/indexedcp/client.py#L57)

Correcting the Server path(CWD)
Server Node: this.outputDir = options.outputDir || process.cwd();
Server Python: upload_dir: str = "./uploads",

Fix: self.upload_dir = Path(upload_dir) if upload_dir else Path.cwd() [LINK](https://github.com/mieweb/indexedcp/blob/05c0d2839a3f7fded6b5a9a406037b6401666bce/python/src/indexedcp/server.py#L41)

horner commented on python/examples/client_demo.py:
As a demonstration. This thing is way too complex. We want to show three or four lines of Python code for a client to work.

adithya1012 replied:
Previously I have included Multiple ways we can upload the file through client_demo.py 
Short : https://www.youtube.com/shorts/RSIwcoRnTu0

I have updated the client_demo.py to 50 lines which includes the core 4 lines.

```
    await client.initialize()
    await client.add_file(demo_file)
    results = await client.upload_buffered_files()
    await client.close()
```

horner commented on python/examples/server_demo.py:
This is good.

adithya1012 replied:
Thank you !

horner replied:
Actually tho:
**NOTE** This is a beta package and encryption is not supported in the client or in the server.

* Server code should removed
* encryption on the client should be supported.

horner commented on python/src/indexedcp/client.py:
```diff
@@ -0,0 +1,613 @@
+
+class IndexedCPClient:
+    
+    def __init__(
```
I don't see any client side encryption

adithya1012 replied:
I have not included the Encryption in this PR (to keep it less code). I am working on it. https://github.com/adithya1012/IndexedCP/issues/22

adithya1012 committed:
PR Comments fix: align with Node.js defaults and simplify docs

- Server upload_dir now defaults to cwd() (matches Node.js)
- Simplify README: focus on usage, remove implementation details
- Simplify client_demo.py: 50 lines vs 380, shows 4-line usage

horner committed:
Added python info to the main readme

adithya1012 committed:
feat(python): Add crypto utilities and keystore system

- Implement RSA-4096 and AES-256-GCM encryption (crypto_utils.py)
- Add pluggable keystore abstraction with filesystem implementation
- Include 44 comprehensive tests (21 crypto + 23 keystore)
- Set secure file permissions (0600/0700) with thread-safe locking

adithya1012 committed:
Merge pull request #44 from adithya1012/python_indexcp-encryption

feat(python): Add crypto utilities and keystore system

adithya1012 committed:
fix(python): Remove python server related.

- Updated examples to support Node server
- Updated integration testing
- Readme updated

adithya1012 committed:
feat: Add EncryptedStorage implementation for IndexedCP.

- Introduced EncryptedStorage class for encrypted storage.
- Updated create_storage function to support encrypted storage.
- Modified FileSystemKeyStore and SQLiteStorage initialization logs for consistency.
- Updated tests to support encryption features in IndexedCPClient.
